tsi: panic: runtime error: index out of range

Running `inch -t 32000000 -time 24h -v -p 10000 -c 5` and then dropping the hot shard caused this panic.

 
```
panic: runtime error: index out of range

goroutine 170148 [running]:
encoding/binary.PutUvarint(0xc55129ce63, 0x5, 0x5, 0x3d09be0e05, 0x4950427)
	/usr/local/go/src/encoding/binary/varint.go:48 +0x61
github.com/influxdata/influxdb/tsdb/index/tsi1.(*TagBlockEncoder).EncodeValue(0xc48a74fe00, 0x7fb80d039bd9, 0xc, 0x4950427, 0xc55cde8500, 0xc574957c30, 0x1, 0x1, 0xc55129cf00, 0xc55129cfc0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/tag_block.go:599 +0x2e1
github.com/influxdata/influxdb/tsdb/index/tsi1.(*LogFile).writeTagsetTo(0xc43e446000, 0x127fdc0, 0xc590434240, 0xc50723dcb8, 0x2, 0xc55129d1d0, 0xc55129d310, 0x0, 0x5000104)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:855 +0x4b4
github.com/influxdata/influxdb/tsdb/index/tsi1.(*LogFile).writeTagsetsTo(0xc43e446000, 0x127fdc0, 0xc590434240, 0xc58a8bd770, 0x1, 0x1, 0xc55129d1d0, 0xc55129d310, 0x8, 0xcd5800)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:823 +0xa7
github.com/influxdata/influxdb/tsdb/index/tsi1.(*LogFile).CompactTo(0xc43e446000, 0x1284200, 0xc437398060, 0x2000000, 0x6, 0x4, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:779 +0x2b3
github.com/influxdata/influxdb/tsdb/index/tsi1.(*Partition).compactLogFile(0xc4202e8d80, 0xc43e446000)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:990 +0x68c
github.com/influxdata/influxdb/tsdb/index/tsi1.(*Partition).checkLogFile.func1(0xc4202e8d80, 0xc43e446000)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:956 +0x68
created by github.com/influxdata/influxdb/tsdb/index/tsi1.(*Partition).checkLogFile
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:954 +0xcc
```

--- a/tsdb/index/tsi1/partition.go
+++ b/tsdb/index/tsi1/partition.go
@@ -296,13 +296,12 @@ func (i *Partition) Wait() {
 // Close closes the index.
 func (i *Partition) Close() error {
 	// Wait for goroutines to finish outstanding compactions.
+	i.once.Do(func() { close(i.closing) })
 	i.wg.Wait()
 
 	// Lock index and close remaining
 	i.mu.Lock()
 	defer i.mu.Unlock()
-
-	i.once.Do(func() { close(i.closing) })
 
 	// Close log files.
 	for _, f := range i.fileSet.files {
@@ -315,14 +314,14 @@ func (i *Partition) Close() error {
 
 // closing returns true if the partition is currently closing. It does not require
 // a lock so will always return to callers.
-// func (i *Partition) closing() bool {
-// 	select {
-// 	case <-i.closing:
-// 		return true
-// 	default:
-// 		return false
-// 	}
-// }
+func (p *Partition) isClosing() bool {
+	select {
+	case <-p.closing:
+		return true
+	default:
+		return false
+	}
+}
 
 // Path returns the path to the partition.
 func (i *Partition) Path() string { return i.path }
@@ -766,7 +765,9 @@ func (i *Partition) Compact() {
 
 // compact compacts continguous groups of files that are not currently compacting.
 func (i *Partition) compact() {
-	if i.compactionsDisabled {
+	if i.isClosing() {
+		return
+	} else if i.compactionsDisabled {
 		return
 	}
 
@@ -964,6 +965,10 @@ func (i *Partition) checkLogFile() error {
 // same identifier but will have a ".tsi" extension. Once the log file is
 // compacted then the manifest is updated and the log file is discarded.
 func (i *Partition) compactLogFile(logFile *LogFile) {
+	if i.isClosing() {
+		return
+	}
+
 	start := time.Now()
 
 	// Retrieve identifier from current path.

